File: python/CodeWars/Nivel 3 prime streaming.py
"""
Create an endless generator that yields prime numbers. the generator must be able to produce a million primes in a few seconds
"""


def generate_primes(n):
    is_prime = [True] * (n + 1) 
    is_prime[0] = is_prime[1] = False 
    
    i = 2
    while i * i <= n:
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
        i += 1

    return [x for x in range(n + 1) if is_prime[x]]

class Primes:
    n = 16000000
    all_primes = generate_primes(n)
    
    @staticmethod
    def stream():
        yield from Primes.all_primes


generate_primes(16000000)

File: CodeWars/Nivel3PrimeStreaming.java
import java.util.ArrayList;
import java.util.List;

public class Nivel3PrimeStreaming {
    private static boolean[] generatePrimes(int n) {
        boolean[] isPrime = new boolean[n + 1];
        for (int i = 2; i <= n; i++) {
            isPrime[i] = true;
        }
        isPrime[0] = isPrime[1] = false;
        
        int i = 2;
        while (i * i <= n) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
            i++;
        }
        
        return isPrime;
    }
    
    private static List<Integer> generatePrimesList(int n) {
        boolean[] isPrime = generatePrimes(n);
        List<Integer> primes = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }
    
    public static List<Integer> stream() {
        int n = 16000000;
        return generatePrimesList(n);
    }
}
